include/pybind11/pytypes.h

```diff
@@ -796,11 +796,16 @@ PYBIND11_NAMESPACE_END(detail)
     : Parent(check_(o) ? o.release().ptr() : ConvertFun(o.ptr()), stolen_t{}) \
     { if (!m_ptr) throw error_already_set(); }
 
+#define PYBIND11_OBJECT_CHECK_FAILED(Name, o) \
+    type_error("Object of type '" + std::string(Py_TYPE(o.ptr())->tp_name) + "' is not an instance of '" #Name "'")
```

Comment:
I'm not sure this works correctly for nested classes. Questions:
* Is there a higher-level function to get the fully-qualified class name?
* Could we exercise with a native nested class, and a nested pybind11 extension class?

Reply:
I know it looks horrible, but `Py_TYPE(o.ptr())->tp_name` is used in other places as well.
About nested types: good point. I should try. Probably something with `__qualname__`.

Reply:
> I know it looks horrible, but `Py_TYPE(o.ptr())->tp_name` is used in other places as well.
> About nested types: good point. I should try. Probably something with `__qualname__`.

This thought randomly crossed my mind:
The "proper" way of doing this is:
* Implement a helper function that returns the fully-qualified type name (module + qualname + tp_name) as a string.
* Unit test that separately.
* Hook it in here, the existing simple test is fine as-is.

Reply:
Yeah, I just need to find time to check how the full name is normally retrieved :-)

Reply:
Actually, a quick lookup confirms that `tp_name` is the way to go?

> Pointer to a NUL-terminated string containing the name of the type. For types that are accessible as module globals, the string should be the full module name, followed by a dot, followed by the type name; for built-in types, it should be just the type name. If the module is a submodule of a package, the full package name is part of the full module name. For example, a type named T defined in module M in subpackage Q in package P should have the tp_name initializer "P.Q.M.T".

https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_name

Let me confirm with a few tests ;-)

Reply:
Fix needed for PyPy, though, it seems. It looks like PyPy has `tp_name` the same as `__name__`. Need to check and confirm if I got that correct with half-sleepy mind.

Reply:
I've merged the `tp_name` PR, should this be rebased?